x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -69,49 +79,127 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
+                        }
+                        // we should make sure that allocating two copies of the index's primary shards to that node doesn't put the node
+                        // above the low watermark, if not, the new shrunken index may not be initialized successfully
+                        long shardsOnCurrentNodeStorageBytes = 0;
+                        if (nodeShardsStorageBytes.containsKey(node.nodeId())) {
+                            shardsOnCurrentNodeStorageBytes = nodeShardsStorageBytes.get(node.nodeId());
+                        }
+                        if (nodeAvailableBytes > freeBytesThresholdLow + 2 * indexPrimaryShardsStorageBytes -
+                            shardsOnCurrentNodeStorageBytes) {
+                            validRoutingNodes.add(node);
+                        }
+                    }
+                }
+
+                for (RoutingNode node : validRoutingNodes) {
+                    boolean canAllocateOneCopyOfEachShard = routingsByShardId.values().stream()
+                        .allMatch(shardRoutings -> shardRoutings.stream()
+                            .map(shardRouting -> allocationDeciders.canAllocate(shardRouting, node, allocation).type())
+                            .anyMatch(Decision.Type.YES::equals));
+                    if (canAllocateOneCopyOfEachShard) {
+                        validNodeIds.add(node.node().getId());
+                    }
+                }
+                if (validNodeIds.size() == 0) {
+                    logger.debug("could not find any nodes to allocate index [{}] onto prior to shrink", indexName);
+                    listener.onFailure(new NoNodeAvailableException("could not find any nodes to allocate index [" +
+                        indexName + "] onto prior to shrink"));
+                    return;
+                }
+
+                List<Map.Entry<String, Long>> nodeShardsStorageList = new ArrayList<>(nodeShardsStorageBytes.entrySet());
+                nodeShardsStorageList.sort((o1, o2) -> o2.getValue().compareTo(o1.getValue()));
+                Optional<String> nodeId = Optional.empty();
+                for (Map.Entry<String, Long> entry : nodeShardsStorageList) {
+                    // we prefer to select the node which contains the maximum shards storage bytes of the index from the valid node list
+                    if (validNodeIds.contains(entry.getKey())) {
+                        nodeId = Optional.of(entry.getKey());
+                        break;
+                    }
+                }
+
+                // if we cannot find a node which contains any shard of the index,
+                // shuffle the valid node list and select randomly
+                if (nodeId.isEmpty()) {
+                    List<String> list = new ArrayList<>(validNodeIds);
+                    Randomness.shuffle(list);
+                    nodeId = list.stream().findAny();
+                }
```

Comment:
I think it'd be useful to factor this into a separate method and then make it unit testable, what do you think?